File: addons/terrabrush/Scripts/CustomContentLoader.cs
#if TOOLS
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Godot;

namespace TerraBrush;

public static class CustomContentLoader {
    public static void AddBrushesPreviewToParent(Node parentNode, Action<int> onSelect, bool useCircleIcon = false) {
        var brushPreviewPrefab = ResourceLoader.Load<PackedScene>("res://addons/terrabrush/Components/DockPreviewButton.tscn");

        var directories = new List<DirAccess>() { DirAccess.Open("res://addons/terrabrush/Assets/Brushes/") };

        var customBrushesDirectory = (string)ProjectSettings.GetSetting(SettingContants.CustomBrushesFolder);
        if (!string.IsNullOrWhiteSpace(customBrushesDirectory) && DirAccess.DirExistsAbsolute(customBrushesDirectory)) {
            directories.Add(DirAccess.Open(customBrushesDirectory));
        }

        var index = 0;
        foreach (var directory in directories) {
            foreach (var file in directory.GetFiles()) {
                if (file.EndsWith(".png", StringComparison.InvariantCultureIgnoreCase)) {
                    var brushImage = ResourceLoader.Load<Texture2D>(Path.Combine(directory.GetCurrentDir(), file));
                    var dockPreviewButton = brushPreviewPrefab.Instantiate<DockPreviewButton>();
                    dockPreviewButton.IconType = useCircleIcon ? IconType.Circle : IconType.Square;
                    dockPreviewButton.Margin = 5;
                    parentNode.AddChild(dockPreviewButton);

                    dockPreviewButton.ButtonImage = brushImage; // We keep the original image in the ButtonImage so we can use it later.

                    var whiteBrushImage = new Image();
                    whiteBrushImage.CopyFrom(brushImage.GetImage());

                    for (var x = 0; x < whiteBrushImage.GetWidth(); x++) {
                        for (var y = 0; y < whiteBrushImage.GetHeight(); y++) {
                            var pixel = whiteBrushImage.GetPixel(x, y);
                            whiteBrushImage.SetPixel(x, y, new Color(1.0f - pixel.R, 1.0f - pixel.G, 1.0f - pixel.B, pixel.A));
                        }
                    }

                    dockPreviewButton.SetTextureImage(ImageTexture.CreateFromImage(whiteBrushImage));

                    var currentIndex = index;
                    dockPreviewButton.OnSelect = () => {
                        onSelect(currentIndex);
                    };

                    index += 1;
                }
            }
        }
    }

    public static void AddTexturesPreviewToParent(TerraBrush terraBrush, Node parentNode, Action<int> onSelect, bool useCircleIcon = false) {
        if (terraBrush.TextureSets?.TextureSets != null) {
            var texturePreviewPrefab = ResourceLoader.Load<PackedScene>("res://addons/terrabrush/Components/DockPreviewButton.tscn");

            for (var i = 0; i < terraBrush.TextureSets.TextureSets.Length; i++) {
                var textureSet = terraBrush.TextureSets.TextureSets[i];

                var dockPreviewButton = texturePreviewPrefab.Instantiate<DockPreviewButton>();
                dockPreviewButton.IconType = useCircleIcon ? IconType.Circle : IconType.Square;
                dockPreviewButton.Margin = 10;
                dockPreviewButton.SetTextureImage(textureSet.AlbedoTexture);
                dockPreviewButton.TooltipText = !string.IsNullOrEmpty(textureSet.Name)
                                                ? textureSet.Name
                                                : $"Texture {i + 1}";

                parentNode.AddChild(dockPreviewButton);


                var currentIndex = i;
                dockPreviewButton.OnSelect = () => {
                    onSelect(currentIndex);
                };
            }
        }
    }

    public static void AddFoliagesPreviewToParent(TerraBrush terraBrush, Node parentNode, Action<int> onSelect, bool useCircleIcon = false) {
        if (terraBrush.Foliages != null) {
            var foliagePreviewPrefab = ResourceLoader.Load<PackedScene>("res://addons/terrabrush/Components/DockPreviewButton.tscn");

            for (var i = 0; i < terraBrush.Foliages.Length; i++) {
                var foliage = terraBrush.Foliages[i];
                if (foliage.Definition != null) {
                    var dockPreviewButton = foliagePreviewPrefab.Instantiate<DockPreviewButton>();
                    dockPreviewButton.IconType = useCircleIcon ? IconType.Circle : IconType.Square;
                    dockPreviewButton.Margin = 5;

                    if (foliage.Definition?.MeshMaterial == null && foliage.Definition?.AlbedoTextures?.Length == 0 && foliage.Definition.Mesh != null) {
                        dockPreviewButton.LoadResourcePreview(foliage.Definition.Mesh);
                    } else if (foliage.Definition?.MeshMaterial != null) {
                        dockPreviewButton.LoadResourcePreview(foliage.Definition.MeshMaterial);
                    } else if (foliage.Definition?.AlbedoTextures?.Length > 0) {
                        dockPreviewButton.LoadResourcePreview(foliage.Definition.AlbedoTextures[0]);
                    }

                    dockPreviewButton.TooltipText = !string.IsNullOrEmpty(foliage.Definition?.Mesh?.ResourcePath)
                                                    ? Path.GetFileName(foliage.Definition.Mesh.ResourcePath)
                                                    : $"Foliage {i + 1}";

                    parentNode.AddChild(dockPreviewButton);

                    var currentIndex = i;
                    dockPreviewButton.OnSelect = () => {
                        onSelect(currentIndex);
                    };
                }
            }
        }
    }

    public static void AddObjectsPreviewToParent(TerraBrush terraBrush, Node parentNode, Action<int> onSelect, bool useCircleIcon = false) {
        if (terraBrush.Objects != null) {
            var objectPreviewPrefab = ResourceLoader.Load<PackedScene>("res://addons/terrabrush/Components/DockPreviewButton.tscn");

            for (var i = 0; i < terraBrush.Objects.Length; i++) {
                var objectItem = terraBrush.Objects[i];

                if (objectItem.Definition?.ObjectScenes?.Length > 0 || objectItem.Definition?.LODMeshes?.Length > 0) {
                    var dockPreviewButton = objectPreviewPrefab.Instantiate<DockPreviewButton>();
                    dockPreviewButton.IconType = useCircleIcon ? IconType.Circle : IconType.Square;
                    dockPreviewButton.Margin = 5;

                    var meshResource = (Resource)(objectItem.Definition?.ObjectScenes?.Length > 0 ? objectItem.Definition?.ObjectScenes[0] : objectItem.Definition?.LODMeshes[0].Meshes[0].Mesh);
                    dockPreviewButton.LoadResourcePreview(meshResource);

                    var packedScenes = objectItem.Definition?.ObjectScenes;
                    var tooltipText = $"Object {i + 1}";
                    if (packedScenes?.Length > 0) {
                        var sceneNames = packedScenes
                            .Where(scene => scene != null)
                            .Select(scene => Path.GetFileName(scene.ResourcePath))
                            .Where(name => !string.IsNullOrEmpty(name));
                        var joinedNames = string.Join(", ", sceneNames);
                        if (!string.IsNullOrEmpty(joinedNames)) {
                            tooltipText = joinedNames;
                        }
                    }

                    dockPreviewButton.TooltipText = tooltipText;

                    parentNode.AddChild(dockPreviewButton);

                    var currentIndex = i;
                    dockPreviewButton.OnSelect = () => {
                        onSelect(currentIndex);
                    };
                }
            }
        }
    }

    public static void AddMetaInfoLayersPreviewToParent(TerraBrush terraBrush, Node parentNode, Action<int> onSelect, bool useCircleIcon = false) {
        if (terraBrush.MetaInfoLayers?.Length > 0) {
            var previewPrefab = ResourceLoader.Load<PackedScene>("res://addons/terrabrush/Components/DockPreviewButton.tscn");

            for (var i = 0; i < terraBrush.MetaInfoLayers.Length; i++) {
                var metaInfoLayer = terraBrush.MetaInfoLayers[i];

                var dockPreviewButton = previewPrefab.Instantiate<DockPreviewButton>();
                dockPreviewButton.IconType = useCircleIcon ? IconType.Circle : IconType.Square;
                dockPreviewButton.Margin = 10;

                var image = Image.CreateEmpty(48, 48, false, Image.Format.Rgba8);
                image.Fill(metaInfoLayer.Color);
                var imageTexture = ImageTexture.CreateFromImage(image);
                dockPreviewButton.SetTextureImage(imageTexture);
                dockPreviewButton.TooltipText = !string.IsNullOrEmpty(metaInfoLayer.Name)
                                                ? metaInfoLayer.Name
                                                : $"MetaInfo {i + 1}";

                parentNode.AddChild(dockPreviewButton);

                var currentIndex = i;
                dockPreviewButton.OnSelect = () => {
                    onSelect(currentIndex);
                };
            }
        }
    }
}
#endif

File: src/misc/custom_content_loader.cpp
#include "custom_content_loader.h"
#include "setting_contants.h"
#include "../editor_nodes/dock_preview_button.h"
#include "../editor_resources/object_definition_resource.h"
#include "../editor_resources/object_resource.h"
#include "../editor_resources/foliage_resource.h"
#include "../editor_resources/foliage_definition_resource.h"
#include "../editor_resources/texture_set_resource.h"
#include "../editor_resources/meta_info_layer_resource.h"

#include <godot_cpp/classes/dir_access.hpp>
#include <godot_cpp/classes/file_access.hpp>
#include <godot_cpp/classes/project_settings.hpp>
#include <godot_cpp/classes/resource_loader.hpp>
#include <godot_cpp/classes/texture2d.hpp>
#include <godot_cpp/classes/image.hpp>
#include <godot_cpp/classes/image_texture.hpp>
#include <godot_cpp/classes/resource.hpp>
#include <godot_cpp/classes/packed_scene.hpp>

using namespace godot;

void CustomContentLoader::addBrushesPreviewToParent(Node *parentNode, Callable onSelect, bool useCircleIcon) {
    TypedArray<Ref<DirAccess>> directories = TypedArray<Ref<DirAccess>>();
    directories.append(DirAccess::open("res://addons/terrabrush/Assets/Brushes/"));

    String customBrushesDirectory = ProjectSettings::get_singleton()->get_setting(SettingContants::CustomBrushesFolder());
    if (!customBrushesDirectory.is_empty() && DirAccess::dir_exists_absolute(customBrushesDirectory)) {
        directories.append(DirAccess::open(customBrushesDirectory));
    }

    int index = 0;
    for (int directoryIndex = 0; directoryIndex < directories.size(); directoryIndex++) {
        Ref<DirAccess> directory = directories[directoryIndex];

        PackedStringArray files = directory->get_files();
        for (int fileIndex = 0; fileIndex < files.size(); fileIndex++) {
            String file = files[fileIndex];

            if (file.ends_with(".png") || file.ends_with(".PNG")) {
                Ref<Texture2D> brushImage = ResourceLoader::get_singleton()->load(directory->get_current_dir() + "/" + file);
                DockPreviewButton *dockPreviewButton = memnew(DockPreviewButton);
                dockPreviewButton->set_iconType(useCircleIcon ? DockPreviewButton::IconType::Circle : DockPreviewButton::IconType::Square);
                dockPreviewButton->set_margin(5);
                parentNode->add_child(dockPreviewButton);

                dockPreviewButton->set_buttonImage(brushImage); // We keep the original image in the ButtonImage so we can use it later.

                Ref<Image> whiteBrushImage = memnew(Image);
                whiteBrushImage->copy_from(brushImage->get_image());

                for (int x = 0; x < whiteBrushImage->get_width(); x++) {
                    for (int y = 0; y < whiteBrushImage->get_height(); y++) {
                        Color pixel = whiteBrushImage->get_pixel(x, y);
                        whiteBrushImage->set_pixel(x, y, Color(1.0f - pixel.r, 1.0f - pixel.g, 1.0f - pixel.b, pixel.a));
                    }
                }

                dockPreviewButton->setTextureImage(ImageTexture::create_from_image(whiteBrushImage));

                int currentIndex = index;
                dockPreviewButton->set_onSelect(onSelect.bind(currentIndex));

                index += 1;
            }
        }
    }
}

void CustomContentLoader::addTexturesPreviewToParent(TerraBrush *terraBrush, Node *parentNode, Callable onSelect, bool useCircleIcon) {
    if (!terraBrush->get_textureSets().is_null() && terraBrush->get_textureSets()->get_textureSets().size() > 0) {
        for (int i = 0; i < terraBrush->get_textureSets()->get_textureSets().size(); i++) {
            Ref<TextureSetResource> textureSet = terraBrush->get_textureSets()->get_textureSets()[i];
            if (textureSet.is_null()) {
                break;
            }

            DockPreviewButton *dockPreviewButton = memnew(DockPreviewButton);

            dockPreviewButton->set_iconType(useCircleIcon ? DockPreviewButton::IconType::Circle : DockPreviewButton::IconType::Square);
            dockPreviewButton->set_margin(10);

            parentNode->add_child(dockPreviewButton);

            dockPreviewButton->setTextureImage(textureSet->get_albedoTexture());
            dockPreviewButton->set_tooltip_text(
                !textureSet->get_name().is_empty()
                    ? textureSet->get_name()
                    : "Texture " + String::num_int64(i + 1)
            );

            int currentIndex = i;
            dockPreviewButton->set_onSelect(onSelect.bind(currentIndex));
        }
    }
}

void CustomContentLoader::addFoliagesPreviewToParent(TerraBrush *terraBrush, Node *parentNode, Callable onSelect, bool useCircleIcon) {
    if (terraBrush->get_foliages().size() > 0) {
        for (int i = 0; i < terraBrush->get_foliages().size(); i++) {
            Ref<FoliageResource> foliage = terraBrush->get_foliages()[i];
            if (!foliage.is_null() && !foliage->get_definition().is_null()) {
                DockPreviewButton *dockPreviewButton = memnew(DockPreviewButton);

                dockPreviewButton->set_iconType(useCircleIcon ? DockPreviewButton::IconType::Circle : DockPreviewButton::IconType::Square);
                dockPreviewButton->set_margin(5);

                parentNode->add_child(dockPreviewButton);

                if (foliage->get_definition()->get_meshMaterial().is_null() && foliage->get_definition()->get_albedoTextures().size() == 0 && !foliage->get_definition()->get_mesh().is_null()) {
                    dockPreviewButton->loadResourcePreview(foliage->get_definition()->get_mesh());
                } else if (!foliage->get_definition()->get_meshMaterial().is_null()) {
                    dockPreviewButton->loadResourcePreview(foliage->get_definition()->get_meshMaterial());
                } else if (foliage->get_definition()->get_albedoTextures().size() > 0) {
                    dockPreviewButton->loadResourcePreview(foliage->get_definition()->get_albedoTextures()[0]);
                }

                dockPreviewButton->set_tooltip_text(
                    !foliage->get_definition()->get_mesh().is_null() && !foliage->get_definition()->get_mesh()->get_path().is_empty()
                        ? foliage->get_definition()->get_mesh()->get_path().get_file()
                        : "Foliage " + String::num_int64(i + 1)
                );

                int currentIndex = i;
                dockPreviewButton->set_onSelect(onSelect.bind(currentIndex));
            }
        }
    }
}

void CustomContentLoader::addObjectsPreviewToParent(TerraBrush *terraBrush, Node *parentNode, Callable onSelect, bool useCircleIcon) {
    if (terraBrush->get_objects().size() > 0) {
        for (int i = 0; i < terraBrush->get_objects().size(); i++) {
            Ref<ObjectResource> objectItem = terraBrush->get_objects()[i];

            if (objectItem.is_null() || objectItem->get_definition().is_null()) {
                break;
            }

            if (objectItem->get_definition()->get_objectScenes().size() > 0 || objectItem->get_definition()->get_lodMeshes().size() > 0) {
                DockPreviewButton *dockPreviewButton = memnew(DockPreviewButton);

                dockPreviewButton->set_iconType(useCircleIcon ? DockPreviewButton::IconType::Circle : DockPreviewButton::IconType::Square);
                dockPreviewButton->set_margin(5);

                parentNode->add_child(dockPreviewButton);

                Ref<Resource> meshResource = objectItem->get_definition()->get_objectScenes().size() > 0 ? objectItem->get_definition()->get_objectScenes()[0] : (Ref<ObjectOctreeLODMeshDefinitionResource>((Ref<ObjectOctreeLODMeshesDefinitionResource>(objectItem->get_definition()->get_lodMeshes()[0]))->get_meshes()[0]))->get_mesh();
                dockPreviewButton->loadResourcePreview(meshResource);

                TypedArray<Ref<PackedScene>> packedScenes = objectItem->get_definition()->get_objectScenes();
                String tooltipText = "Object " + String::num_int64(i + 1);
                if (packedScenes.size() > 0) {
                    PackedStringArray sceneNames = TypedArray<String>();
                    for (Ref<PackedScene> packedScene : packedScenes) {
                        if (!packedScene.is_null() && !packedScene->get_path().get_file().is_empty()) {
                            sceneNames.append(packedScene->get_path().get_file());
                        }
                    }

                    String joinedNames = String(",").join(sceneNames);
                    if (!joinedNames.is_empty()) {
                        tooltipText = joinedNames;
                    }
                }

                dockPreviewButton->set_tooltip_text(tooltipText);

                int currentIndex = i;
                dockPreviewButton->set_onSelect(onSelect.bind(currentIndex));
            }
        }
    }
}

void CustomContentLoader::addMetaInfoLayersPreviewToParent(TerraBrush *terraBrush, Node *parentNode, Callable onSelect, bool useCircleIcon) {
    if (terraBrush->get_metaInfoLayers().size() > 0) {
        for (int i = 0; i < terraBrush->get_metaInfoLayers().size(); i++) {
            Ref<MetaInfoLayerResource> metaInfoLayer = terraBrush->get_metaInfoLayers()[i];
            if (metaInfoLayer.is_null()) {
                break;
            }

            DockPreviewButton *dockPreviewButton = memnew(DockPreviewButton);

            dockPreviewButton->set_iconType(useCircleIcon ? DockPreviewButton::IconType::Circle : DockPreviewButton::IconType::Square);
            dockPreviewButton->set_margin(10);

            parentNode->add_child(dockPreviewButton);

            Ref<Image> image = Image::create_empty(48, 48, false, Image::Format::FORMAT_RGBA8);
            image->fill(metaInfoLayer->get_color());
            Ref<ImageTexture> imageTexture = ImageTexture::create_from_image(image);
            dockPreviewButton->setTextureImage(imageTexture);
            dockPreviewButton->set_tooltip_text(
                !metaInfoLayer->get_name()
                    ? metaInfoLayer->get_name()
                    : "MetaInfo " + String::num_int64(i + 1)
            );

            int currentIndex = i;
            dockPreviewButton->set_onSelect(onSelect.bind(currentIndex));
        }
    }
}